units/unexpirable_spawnhost_stats.go

```diff
@@ -0,0 +1,115 @@
+package units
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/evergreen-ci/evergreen"
+	"github.com/evergreen-ci/evergreen/model/host"
+	"github.com/evergreen-ci/utility"
+	"github.com/mongodb/amboy"
+	"github.com/mongodb/amboy/job"
+	"github.com/mongodb/amboy/registry"
+	"github.com/mongodb/grip"
+	"github.com/mongodb/grip/message"
+	"github.com/pkg/errors"
+)
+
+const (
+	unexpirableSpawnHostStatsJobName        = "unexpirable-spawnhost-stats"
+	unexpirableSpawnHostStatsJobMaxAttempts = 10
+)
+
+func init() {
+	registry.AddJobType(unexpirableSpawnHostStatsJobName, func() amboy.Job {
+		return makeUnexpirableSpawnHostStatsJob()
+	})
+}
+
+type unexpirableSpawnHostStatsJob struct {
+	job.Base `bson:"metadata" json:"metadata" yaml:"metadata"`
+}
+
+func makeUnexpirableSpawnHostStatsJob() *unexpirableSpawnHostStatsJob {
+	j := &unexpirableSpawnHostStatsJob{
+		Base: job.Base{
+			JobType: amboy.JobType{
+				Name:    unexpirableSpawnHostStatsJobName,
+				Version: 0,
+			},
+		},
+	}
+	return j
+}
+
+// NewUnexpirableSpawnHostStatsJob returns a job to collect estimated statistics
+// on unexpirable spawn host usage.
+func NewUnexpirableSpawnHostStatsJob(ts string) amboy.Job {
+	j := makeUnexpirableSpawnHostStatsJob()
+	j.SetID(fmt.Sprintf("%s.%s", unexpirableSpawnHostStatsJobName, ts))
+	j.SetScopes([]string{unexpirableSpawnHostStatsJobName})
+	j.SetEnqueueAllScopes(true)
+	j.UpdateRetryInfo(amboy.JobRetryOptions{
+		Retryable:   utility.TruePtr(),
+		MaxAttempts: utility.ToIntPtr(unexpirableSpawnHostStatsJobMaxAttempts),
+	})
+	return j
+}
+
+func (j *unexpirableSpawnHostStatsJob) Run(ctx context.Context) {
+	defer j.MarkComplete()
+
+	hosts, err := host.FindUnexpirableRunning()
+	if err != nil {
+		j.AddRetryableError(errors.Wrap(err, "finding unexpirable running hosts"))
+		return
+	}
+
+	stats := j.getStats(hosts)
+
+	grip.Info(message.Fields{
```

Comment:
In the ticket we also mentioned logging how many hosts are stopped, is that no longer something we want to do?

Reply:
So I did think about that, but for the sleep schedule, it would probably matter more to distinguish why the host is stopped (whether it's for the nightly sleep schedule vs. whether it's off 24/7). For now, it only stops if the user manually stops it, so there's not much point in just logging the number of stopped hosts until the sleep schedule actually exists.

The downtime isn't greatly important. The uptime is the more important measure just because we need a relative comparison of current uptime vs. uptime after savings measures.